iodata/inputs/gaussian.py

```diff
@@ -0,0 +1,75 @@
+# IODATA is an input and output module for quantum chemistry.
+# Copyright (C) 2011-2019 The IODATA Development Team
+#
+# This file is part of IODATA.
+#
+# IODATA is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 3
+# of the License, or (at your option) any later version.
+#
+# IODATA is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, see <http://www.gnu.org/licenses/>
+# --
+"""Gaussian Input Module."""
+
+
+# from attr import asdict
+from typing import TextIO
+from string import Template
+
+from ..iodata import IOData
+from ..periodic import num2sym
+from ..utils import angstrom
+
+
+__all__ = []
+
+
+default_template = """\
+#n ${lot}/${obasis_name} ${run_type}
+
+${title}
+
+${charge} ${spinmult}
+${geometry}
+
+"""
+
+
+def write_input(f: TextIO, data: IOData, template: str = None):
+
+    # load IOData dict using attr.asdict because the IOData class uses __slots__
+    # fields = asdict(data)
```

Comment:
Using this to convert an instance of `IOData` to dictionary failed with `TypeError: unhashable type: 'list'` for me, so I commented it out and constructed `fields` dictionary manually for the mean while.

Reply:
An example that rise this error:
`>>> from iodata import load_one, write_input
>>> mol = load_one("iodata/test/data/nh3_orca.molden")
>>> write_input(mol, "/Users/lmacaya/Desktop/test_orca.inp", fmt='orca', template="iodata/test/data/template_orca.com")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/lmacaya/Desktop/horton3/write_input/iodata/iodata/api.py", line 252, in write_input
    input_module.write_input(f, iodata, template=template, **kwargs)
  File "/Users/lmacaya/Desktop/horton3/write_input/iodata/iodata/inputs/orca.py", line 43, in write_input
    fields = attr.asdict(data)
  File "/Users/lmacaya/miniconda3/envs/iodata-dev-python-3.7/lib/python3.7/site-packages/attr/_funcs.py", line 62, in asdict
    for i in v
  File "/Users/lmacaya/miniconda3/envs/iodata-dev-python-3.7/lib/python3.7/site-packages/attr/_funcs.py", line 62, in <listcomp>
    for i in v
  File "/Users/lmacaya/miniconda3/envs/iodata-dev-python-3.7/lib/python3.7/site-packages/attr/_funcs.py", line 109, in _asdict_anything
    for kk, vv in iteritems(val)
TypeError: unhashable type: 'list'`

If i change the obasis attribute to None, it works without issues.
>>> mol.obasis = None
>>> write_input(mol, "/Users/lmacaya/Desktop/test_orca.inp", fmt='orca', template="iodata/test/data/template_orca.com")`

Reply:
An example that rise this error:
``` 
>>> from iodata import load_one, write_input
>>> mol = load_one("iodata/test/data/nh3_orca.molden")
>>> write_input(mol, "/Users/lmacaya/Desktop/test_orca.inp", fmt='orca', template="iodata/test/data/template_orca.com")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/lmacaya/Desktop/horton3/write_input/iodata/iodata/api.py", line 252, in write_input
    input_module.write_input(f, iodata, template=template, **kwargs)
  File "/Users/lmacaya/Desktop/horton3/write_input/iodata/iodata/inputs/orca.py", line 43, in write_input
    fields = attr.asdict(data)
  File "/Users/lmacaya/miniconda3/envs/iodata-dev-python-3.7/lib/python3.7/site-packages/attr/_funcs.py", line 62, in asdict
    for i in v
  File "/Users/lmacaya/miniconda3/envs/iodata-dev-python-3.7/lib/python3.7/site-packages/attr/_funcs.py", line 62, in <listcomp>
    for i in v
  File "/Users/lmacaya/miniconda3/envs/iodata-dev-python-3.7/lib/python3.7/site-packages/attr/_funcs.py", line 109, in _asdict_anything
    for kk, vv in iteritems(val)
TypeError: unhashable type: 'list'
```

If i change the obasis attribute to None, it works without issues.
```
>>> mol.obasis = None
>>> write_input(mol, "/Users/lmacaya/Desktop/test_orca.inp", fmt='orca', template="iodata/test/data/template_orca.com")```

Reply:
Thanks! I'll take a closer look.

Reply:
It is related to the fact that `attr.asdict` tries to work recursively, and then encounters stuff it cannot handle. I expect this would be fixed when all namedtuples in our code are replaced by attr classes. For now, you can avoid recursion as follows:

```python
from iodata import load_one
import attr
mol = load_one("iodata/test/data/nh3_orca.molden")
attr.asdict(mol, recurse=False)
```

Reply:
Thanks for the workaround @tovrstra. I fixed `gaussian.py`. @lmacaya, I also add `recurse=False` to `orca.py` module to remove the temporary fix (`mol.obasis = None`) in the test module.

Reply:
I wasn't sure exactly how `attrs` works with all of the components of IOData, so that's why I tried `asdict`, definitely didn't know about the recursion. I opened an issue to request a small overview of how we're using `attrs` (#201)

Reply:
Just for reference: it is a known issue of attrs, see https://github.com/python-attrs/attrs/issues/646. The proper solution seems to be the `retain_collection_types=True` option.